### ISSUE TITLE
Support TypeScript@5.2 "using" declarations

### ISSUE THREAD

**Issue by eddyw** [oven-sh/bun#4448]
### What is the problem this feature would solve?

"ES Explicit Resource Management" is stage 3. It works similarly to `defer` in Go or `using` in C#. Allows to define a clean up method which is run before the `return` of a function.

```ts
// From TS Blog Post
function loggy(id: string): Disposable {
    console.log(`Creating ${id}`);

    return {
        [Symbol.dispose]() {
            console.log(`Disposing ${id}`);
        }
    }
}

function func() {
    using a = loggy("a");
    using b = loggy("b");
    {
        using c = loggy("c");
        using d = loggy("d");
    }
    using e = loggy("e");
    return;

    // Unreachable.
    // Never created, never disposed.
    using f = loggy("f");
}

func();
// Creating a
// Creating b
// Creating c
// Creating d
// Disposing d
// Disposing c
// Creating e
// Disposing e
// Disposing b
// Disposing a
```

### What is the feature you are proposing to solve the problem?

It's been implemented on TypeScript as of version 5.2+ as well as in esbuild v0.18.7. It'd be great to have this in Bun.js.

1. [**using** Declarations and Explicit Resource Management](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#using-declarations-and-explicit-resource-management)
2. [Esbuild v0.18.7](https://github.com/evanw/esbuild/releases/tag/v0.18.7)
3. [TC39 - ECMAScript Explicit Resource Management](https://github.com/tc39/proposal-explicit-resource-management)

Opening this issue mostly to track when this will be implemented in Bun.js. It currently doesn't recognize the keyword `using`.

**Comment by TiBianMod** [comment]
+1

**Comment by sirenkovladd** [comment]
+1

**Comment by TiBianMod** [comment]
@Electroid we have any news on this?

**Comment by paperclover** [comment]
I have this almost implemented in #8151. There are some bugs preventing it from being shipped.
